src/commonMain/kotlin/exchange.dydx.abacus/state/manager/V4StateManagerAdaptor.kt

```diff
@@ -1063,24 +1066,39 @@ class V4StateManagerAdaptor(
         params: Map<String, String>?,
         headers: Map<String, String>?,
         private: Boolean,
-        callback: (String?, Int) -> Unit
+        callback: (url: String, response: String?, code: Int) -> Unit
     ) {
-        super.get(url, params, headers, private) { response, httpCode ->
+        super.get(url, params, headers, private) {url, response, httpCode ->
             when (httpCode) {
                 403 -> {
-                    if (indexerRestriction != UsageRestriction.http403Restriction) {
-                        ioImplementations.threading?.async(ThreadingType.abacus) {
-                            indexerRestriction = UsageRestriction.http403Restriction
+                    indexerRestriction = if (response != null) {
+                        val json = Json.parseToJsonElement(response).jsonObject.toMap()
+                        val errors = parser.asList(parser.value(json, "errors"))
+                        val userRestriction = errors?.firstOrNull { error ->
+                            val msg = parser.asString(parser.value(error, "msg"))
+                            msg?.contains("address") == true
                         }
-                    }
+
+                        if (userRestriction !== null)
+                            UsageRestriction.userRestriction
+                        else
+                            UsageRestriction.http403Restriction
+                    } else UsageRestriction.http403Restriction
                 }
 
                 429 -> {
-                    /* We may need better handling of 429 */
-                    callback(response, httpCode)
+                    // retry after 5 seconds
+                    val timer = ioImplementations.timer ?: CoroutineTimer.instance
```

Comment:
Optimization:  Might want to skip the timer run if there is already one pending... i.e.,

`if (restRetryTimers[url] == null) { ... }`